cpu/atmega_common/periph/rtt.c

```diff
@@ -0,0 +1,314 @@
+/*
+ * Copyright (C) 2018 Acutam Automation, LLC
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/*
+ * @ingroup cpu_atmega_common
+ * @ingroup drivers_periph_rtt
+ * @{
+ *
+ * @file
+ * @brief       Low-level ATmega RTT driver implementation
+ *
+ * @author      Matthew Blue <matthew.blue.neuro@gmail.com>
+ *
+ * In order to safely sleep when using the RTT:
+ * 1. Disable interrupts
+ * 2. Write to one of the asynch registers (e.g. TCCR2A)
+ * 3. Wait for ASSR register's busy flags to clear
+ * 4. Re-enable interrupts
+ * 5. Sleep before interrupt re-enable takes effect
+ *
+ * @}
+ */
+
+#include <avr/interrupt.h>
+
+#include "cpu.h"
+#include "irq.h"
+#include "periph/rtt.h"
+#include "periph_conf.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+typedef struct {
+    uint16_t ext_comp;          /* Extend compare to 24-bits */
+    rtt_cb_t alarm_cb;          /* callback called from RTT alarm */
+    void *alarm_arg;            /* argument passed to the callback */
+    rtt_cb_t overflow_cb;       /* callback called when RTT overflows */
+    void *overflow_arg;         /* argument passed to the callback */
+} rtt_state_t;
+
+static rtt_state_t rtt_state;
+static uint16_t ext_cnt;
+
+static inline void _asynch_wait(void)
+{
+    /* Wait until all busy flags clear. According to the datasheet,
+     * this can take up to 2 positive edges of TOSC1 (32kHz). */
+    while (ASSR & ((1 << TCN2UB) | (1 << OCR2AUB) | (1 << OCR2BUB)
+                | (1 << TCR2AUB) | (1 << TCR2BUB))) {}
+}
+
+/* interrupts are disabled here */
+static uint32_t _safe_cnt_get(void)
+{
+    uint8_t cnt = TCNT2;
+
+    /* if an overflow occured since we disabled interrupts, manually
+     * increment ext_cnt
+     */
+    if (TIFR2 & (1 << TOV2)) {
+        ++ext_cnt;
+    }
+
+    return (ext_cnt << 8) | cnt;
+}
```

Comment:
The IRQ flag should is not cleared by this (unless I'm to  So the ISR will still be served and `ext_cnt` ends up being incremented twice, right?

Also, as `cnt` is read before the IRQ flag is checked. It could happen that `TCNT2` has the value of 255 when reading it into `cnt`. But when checking `TIFR2` for the IRQ flag, it might have just overflown. In that case `cnt` should be reset to 0, as otherwise the returned counter would be 255 to high, right?

Btw: Huge :+1: for documenting that interrupts must be disabled there.

Reply:
In the ATmega328P doc it says that `TIFR2` is R/W (for the non-reserved bits). So clearing the IRQ flag should (hopefully on all other ATmegas as well) be as simple as just clearing the corresponding bit in the register.